src/function/aggregate/nested/list.cpp

```diff
@@ -5,49 +5,665 @@
+	} else if (input.GetType().id() == LogicalTypeId::STRUCT) {
+		auto &children = StructVector::GetEntries(input);
+		return CreateStructSegment(allocator, owning_vector, capacity, children);
+	}
+	return CreatePrimitiveSegment(allocator, owning_vector, capacity, input.GetType());
+}
+
+ListSegment *ListFun::GetSegment(Allocator &allocator, vector<unique_ptr<AllocatedData>> &owning_vector,
+                                 LinkedList *linked_list, Vector &input) {
+
+	ListSegment *segment = nullptr;
+
+	// determine segment
+	if (!linked_list->last_segment) {
+		// empty linked list, create the first (and last) segment
+		auto capacity = GetCapacityForNewSegment(linked_list);
+		segment = CreateSegment(allocator, owning_vector, capacity, input);
+		linked_list->first_segment = segment;
+		linked_list->last_segment = segment;
+
+	} else if (linked_list->last_segment->capacity == linked_list->last_segment->count) {
+		// the last segment of the linked list is full, create a new one and append it
+		auto capacity = GetCapacityForNewSegment(linked_list);
+		segment = CreateSegment(allocator, owning_vector, capacity, input);
+		linked_list->last_segment->next = segment;
+		linked_list->last_segment = segment;
+
+	} else {
+		// the last segment of the linked list is not full, append the data to it
+		segment = linked_list->last_segment;
+	}
+
+	D_ASSERT(segment);
+	return segment;
+}
+
+ListSegment *ListFun::GetCharSegment(Allocator &allocator, vector<unique_ptr<AllocatedData>> &owning_vector,
+                                     LinkedList *linked_list) {
+
+	ListSegment *segment = nullptr;
+
+	// determine segment
+	if (!linked_list->last_segment) {
+		// empty linked list, create the first (and last) char segment
+		auto capacity = GetCapacityForNewSegment(linked_list);
+		segment = CreatePrimitiveSegment(allocator, owning_vector, capacity, LogicalTypeId::VARCHAR);
+		linked_list->first_segment = segment;
+		linked_list->last_segment = segment;
+
+	} else if (linked_list->last_segment->capacity == linked_list->last_segment->count) {
+		// the last segment of the linked list is full, create a new char segment and append it
+		auto capacity = GetCapacityForNewSegment(linked_list);
+		segment = CreatePrimitiveSegment(allocator, owning_vector, capacity, LogicalTypeId::VARCHAR);
+		linked_list->last_segment->next = segment;
+		linked_list->last_segment = segment;
+
+	} else {
+		// the last segment of the linked list is not full, append the data to it
+		segment = linked_list->last_segment;
+	}
+
+	D_ASSERT(segment);
+	return segment;
+}
+
+void ListFun::WriteDataToSegment(Allocator &allocator, vector<unique_ptr<AllocatedData>> &owning_vector,
+                                 ListSegment *segment, Vector &input, idx_t &entry_idx, idx_t &count) {
+
+	// get the vector data and the source index of the entry that we want to write
+	UnifiedVectorFormat input_data;
+	input.ToUnifiedFormat(count, input_data);
```

Comment:
I would just flatten everything, calling `ToUnifiedFormat` multiple times for each row is quite complicated/slow